src/VisualStudio/Core/Def/ProjectSystem/OpenTextBufferProvider.cs

```diff
@@ -0,0 +1,313 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.VisualStudio.Editor;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.TextManager.Interop;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
+{
+    /// <summary>
+    /// A class that provides access to the currently open list of files in Visual Studio. You are able to ask for the text buffer for a document on any thread; events are raised
+    /// on thread, and any method that provides a <see cref="IVsHierarchy"/> must be used on the UI thread.
+    /// </summary>
+    [Export(typeof(OpenTextBufferProvider))]
+    internal sealed class OpenTextBufferProvider : IVsRunningDocTableEvents3, IDisposable
+    {
+        private bool _isDisposed = false;
+
+        /// <summary>
+        /// A simple object for asserting when we're on the UI thread.
+        /// </summary>
+        private readonly ForegroundThreadAffinitizedObject _foregroundAffinitization;
+        private readonly IVsEditorAdaptersFactoryService _editorAdaptersFactoryService;
+        private readonly IVsRunningDocumentTable4 _runningDocumentTable;
+
+        private ImmutableArray<IOpenTextBufferEventListener> _listeners = ImmutableArray<IOpenTextBufferEventListener>.Empty;
+
+        /// <summary>
+        /// The map from monikers to open text buffers; because we can only fetch the text buffer on the UI thread, all updates to this must be done from the UI thread.
+        /// </summary>
+        private ImmutableDictionary<string, ITextBuffer> _monikerToTextBufferMap = ImmutableDictionary<string, ITextBuffer>.Empty.WithComparers(StringComparer.OrdinalIgnoreCase);
+
+        private uint _runningDocumentTableEventsCookie;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public OpenTextBufferProvider(
+            IThreadingContext threadingContext,
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider,
+            IAsynchronousOperationListenerProvider listenerProvider)
+        {
+            _foregroundAffinitization = new ForegroundThreadAffinitizedObject(threadingContext, assertIsForeground: false);
```

Comment:
should this inherit from it?

Reply:
That's how we used to use that type, but I've kinda gotten away from that as a favoring of composition over inheritance. Especially today since the various Assert* methods are public so you end up polluting the public surface area of this new type.